### ISSUE TITLE
Use async RPC from qmanager to resource 

### ISSUE THREAD

**Issue by dongahn** [flux-framework/flux-sched#577]
This was discussed before but I couldn't find an existing issue. The expectation is this will increase performance by allowing overlapping between qmanager and resource operations. https://github.com/flux-framework/flux-sched/issues/573#issuecomment-578420286

**Comment by garlick** [comment]
FYI: the libschedutil code pre-fetches jobspec for each inbound alloc request asynchronously, as it is received, if the reactor loop is not blocked.   The scheduler would get alloc callbacks for each alloc request, after the jobspec is fetched.

If the loop is blocked while N alloc requests are received, then upon unblocking, libschedutil sends a burst of N request messages to the KVS.  Each time a response is received, it calls the scheduler's alloc callback. 

**Possible simple, short term fix**
If a scheduling pass consists of multiple, back to back synchronous resource queries, let the reactor make some progress between them.  I _think_ you could call something like
```c
if (flux_reactor_run (flux_get_reactor (h), FLUX_REACTOR_NOWAIT) < 0)
    // handle fatal error
```
You should be prepared to handle alloc callbacks during this call (so use a flag to make sure those jobs just go in the queue and do not trigger reentry)

**Comment by dongahn** [comment]
> FYI: the libschedutil code pre-fetches jobspec for each inbound alloc request asynchronously, as it is received, if the reactor loop is not blocked. The scheduler would get alloc callbacks for each alloc request, after the jobspec is fetched.

> If the loop is blocked while N alloc requests are received, then upon unblocking, libschedutil sends a burst of N request messages to the KVS. Each time a response is received, it calls the scheduler's alloc callback.

This is good to know. So it sounds like the "delayed scheduling" optimization is already done at the libschedutil level! We won't need Issue #575 implemented at the scheduler level then.

**Comment by dongahn** [comment]
> If a scheduling pass consists of multiple, back to back synchronous resource queries, let the reactor make some progress between them. I think you could call something like

One issue here is, though, the code where our queuing policies are implemented is flux-core agnostic. Like [here](https://github.com/flux-framework/flux-sched/blob/master/qmanager/policies/queue_policy_bf_base_impl.hpp#L143) in the case of backfill. The thought was this would provide a better separation of concerns and allow us to have a standalone  command tester for the queue layer. (Similar to `resource-query` for the resource module). 

One thing I am convinced though, whether or not yielding CPU for `libschedutil` is bad for performance. While the scheduler is already doing an expensive backfill iteration, it may be good not to get new alloc requests anyway?

> If the loop is blocked while N alloc requests are received, then upon unblocking, libschedutil sends a burst of N request messages to the KVS. Each time a response is received, it calls the scheduler's alloc callback.

I suspect this may provide an one more optimization opportunity. Right now we call our schedule loop per each alloc request. Maybe we can make it such that we treat the burst of N requests as a single job event so that we run scheduling loop only once at the final job in this burst? From the scheduler point of view, it can't say which alloc request is the last one of this burst. But maybe with some protocol change, this info can be provided from `libschedutil`?

**Comment by garlick** [comment]
> This is good to know. So it sounds like the "delayed scheduling" optimization is already done at the libschedutil level! We won't need Issue #575 implemented at the scheduler level then.

I don't think this is true.  After leaving the scheduling loop and giving control back to the reactor, qmanager still gets a series of disjoint alloc callbacks.  There will be a higher latency now because the work of fetching jobspec has to happen on the critical path instead of concurrent with the scheduling loop, but without somehow batching allocs (and frees) up before calling the scheduling loop, the problem in #575 persists.

I think you were correct in opening two performance bugs:
- this one:  the synchronous RPCs in qmanager limit concurrency
- #575:  the scheduling loop is run once per alloc request

**Comment by dongahn** [comment]
> #575: the scheduling loop is run once per alloc request

 Presumably we can use prep/check/idle watchers at the `qmanager` level to achieve this. @garlick: does `libschedutil` do anything that would prevent its user (qmanager) to implement these watchers?

**Comment by dongahn** [comment]
> Edit: which is not to say libschedutil wouldn't be able to help with #575 - maybe it could queue requests internally until some high water mark or timeout is reached before making the callback?

As far as the `alloc` request is *individually* sent to qmanager through the callback, I don't think this will be helpful, though. Maybe what we need is another callback type to demarcate the end of bulk submit (or timeout). Then, `qmanager` will just queue up at the `alloc` request handler and call the schedule loop on that new callback. 

Maybe that's sort of what you are proposing @garlick?

**Comment by garlick** [comment]
> Then, qmanager will just queue up at the alloc request handler and call the schedule loop on that new callback

Yes sorry, I should have said that it would be a new callback type. 

>Presumably we can use prep/check/idle watchers at the qmanager level to achieve this. @garlick: does libschedutil do anything that would prevent its user (qmanager) to implement these watchers?

That's how I originally imagined this would be handled, i.e. that the alloc/free callbacks would update a queue belonging to the scheduler, and the scheduler would decide when to call the scheduling loop.

However, it seems like it may be better for libschedutil to build its own queue, let sched pull from that in batches.  That would allow you to address the issue of this bug (the blocking RPCs limiting concurrency) without having the qmanager's queue modified during the course of the scheduling loop.

Does that make sense?

**Comment by dongahn** [comment]
> That's how I originally imagined this would be handled, i.e. that the alloc/free callbacks would update a queue belonging to the scheduler, and the scheduler would decide when to call the scheduling loop.

Great! And I like this idea a lot. 

> However, it seems like it may be better for libschedutil to build its own queue, and let sched pull from that in batches.

Well, at first glance I am a bit worried about introducing yet another queue though. We already have a queue in job-manager and we have multiple queues in qmanager and this will add yet one more interim queue. I implemented the queues in qmanager very efficiently so I think it would be wise to rely on that queue as much as possible.

>  That would allow you to address the issue of this bug (the blocking RPCs limiting concurrency) without having the qmanager's queue modified during the course of the scheduling loop.

Here is my $0.02 for this. When we change the calling of schedule loop on the newly proposed callback type, the impact of the blocking RPCs will be reduced drastically in particular for bulk submit case. 

Then, I was going to look into how I can use async RPCs between qmanager and resource. I am sure there would be a good solution. It just that I haven't looked into that yet. When combined this with the new type of callback, I have to believe we can address both issues without adding more complexity...

**Comment by garlick** [comment]
That was slightly confusing :-) .  So to summarize, it sounds like we don't need to change anything in libschedudtil (because we don't want it to re-implement the queue), and you will look at solving #575 first and then revisit this issue to decide its priority?

**Comment by dongahn** [comment]
>  So to summarize, it sounds like we don't need to change anything in libschedudtil (because we don't want it to re-implement the queue),

Sorry. I was actually suggesting to implement the prep/check/idle watcher scheme (if we don't already have it) within `libschedutil` which invokes a new callback "burst-done". This way, qmanager will only enqueue a job when each alloc request comes and runs its schedule loop when the burst-done callback is called? 

I may misunderstand how `libschedutil` works. But I was envisioning `libschedutil` wouldn't need an additional queue for this...

>  you will look at solving #575 first

Yes.

**Comment by garlick** [comment]
As far as I understand, libschedutil would need to put new jobs _somewhere_ while waiting for enough to accumulate to send them on to the scheduler.  Maybe the job manager could do some batching of alloc requests though.  Let me ponder that one a bit...
